drivers/lcd/lcd.c

```diff
@@ -341,12 +372,35 @@ int lcd_init(lcd_t *dev, const lcd_params_t *params)
         }
     }
     else {
-#endif
+#endif /* IS_USED(MODULE_LCD_SPI) */
+
 #if IS_USED(MODULE_LCD_PARALLEL)
-        lcd_ll_par_gpio_init(dev);
-#else
+        mutex_init(&dev->lock);
+#if IS_USED(MODULE_LCD_PARALLEL_16BIT)
+        dev->word_access = false;
+#endif /* IS_USED(MODULE_LCD_PARALLEL_16BIT) */
+
+        /* Low-level parallel interface initialization */
+        if (lcd_ll_par_driver.init) {
+            lcd_ll_par_driver.init(dev);
+        }
+        else {
+           lcd_ll_par_gpio_init(dev);
+        }
```

Comment:
Can we have a `lcd_parallel_gpio` or `lcd_parallel_periph` pseudo-module so we can have this check compile-time const if only one of the two is used - and don't have to include all the bit-banging code when a periph implementation is used.

Reply:
Makes sense.

Reply:
actually, why have the switch at all instead of 

```C
lcd_ll_par_driver.init = lcd_ll_par_gpio_init;
```

for the bit-banging case

Reply:
It is not necessary that the MCU implements all low-level functions. For example, the ESP32 implementation doesn't define the `init` and `cmd_start` driver functions. If the MCU-driven low-level implementation doesn't define them, the corresponding `lcd_ll_par_gpio_*` functions are used instead.

Alternatively, we could set those members of the driver structure to the `lcd_ll_par_gpio_*` functions in the MCU-driven low-level implementation.

I'm still not really happy having only one low-level driver instance. This avoids to use multiple displays with parallel interface but different low-level implementations. Ideally, each display device should point to it's own low-level implementation. But the configuration via the constant LCD display parameter set seems to be too complicated. Doing it in RAM during the initialization of the device would cost a lot of RAM.

Reply:
On the other hand, the question arises whether there will ever be a board with two or more displays and whether multiple displays with different low-level implementations would then be used. Parallel interfaces occupy soooooo many GPIOs that we can probably assume that if we used the parallel interface, all displays would use the same interface but with different select signals. If this assumption is true, only one low-level driver instance is sufficient.